docs/IMPLEMENTATION.md

```diff
@@ -0,0 +1,207 @@
+
+### Step 5: Test your build system
+
+```bash
+yarn map-gl-style-build \
+  --style-dir=templates/styles \
+  --layer-dir=templates/layers \
+  --out-dir=build
+```
+
+Compare the generated stylesheets in `build/` with your original files to ensure they match.
+
+### Troubleshooting Conversion
+
+**Common issues:**
+- **Missing layers**: Check that all layers from your original styles are present in the templates
+- **Incorrect styling**: Verify that layer IDs, types, and sources match between variants
+- **Missing sources**: Ensure all data sources are defined in your style templates
+
+## Building From Scratch
+
+If you're starting with no existing stylesheets, this section explains how to create the template structure manually.
+
+### Project Structure
+
+The build system organizes layer definitions as individual template files, making them easier to edit directly, track in source control, or use with external style editors. You can also define variables and give them different values by creating style variants.
+
+A simple build command exports a stylesheet for each variant, which you can reference in your Maplibre or Mapbox map.
+
+### Layer Templates
+
+The foundation of this system is the layer template - a JavaScript module that defines how a single map layer should be styled. Each layer template exports a function that receives a `context` object containing variables and returns two objects: `baseStyle` and `overrides`.
+
+#### Basic Structure
+
+```javascript
+// templates/layers/roads.js
+module.exports = (context) => {
+  const baseStyle = {
+    id: 'roads',
+    type: 'line',
+    source: context.roadSource,
+    paint: {
+      'line-color': context.theme.colors.roads,
+      'line-width': context.layout.iconSize
+    }
+  };
+
+  const overrides = {};
+  
+  // Conditional styling based on context
+  if (context.styleName === 'dark-theme') {
+    overrides.paint = {
+      'line-color': context.theme.colors.roads
+    };
+  }
+
+  return { baseStyle, overrides };
+};
+```
+
+### Styles
+
+Once you have a set of layer templates, you can organize them into a complete map style, or a set of related map styles. You create style files manually in the `templates/styles/` directory - these define the context (variables) and layer order for each style variant. 
+
+#### Required Structure
+
+Styles define the context (variables) and layer order. Each variant file must export two objects: `context` and `template`. Every style file must follow this structure:
+
+```javascript
+// templates/styles/My_Style.js
+module.exports.context = {
+  // Variables available to layers
+  theme: { /* ... */ },
+  layout: { /* ... */ }
+};
+
+module.exports.template = {
+  version: 8,
+  name: "My Style Name",
+  sources: { /* ... */ },
+  layers: [ /* ... */ ]
+};
+```
+
+#### Context Requirements
+
+The `context` object contains variables that layers can reference. Each variable file must be declared using `require()`:
```

Comment:
I think we need to be more specific about this being a workflow someone could choose. You do not need external variable files and they don't need to follow a specific format.

We have a format we use that our helper scripts are based on assisting with, but users could do a lot of different things here.

Also, I need to double check, but in theory I think a user could make these module files and use `import` vs `require` so not sure we want to specify that necessarily.